### PR TITLE
added instruction set for CPU 6502

### PR DESCRIPTION
**Pull request by KOomaXX**
I've repushed using rustfmt and making the changes you asked  for. I'm not really happpy for instructions file formatting now. It has turned out to be way longer than previous version. Maybe each set should be moved in a separate file ?

### CODE REVIEW

File: src/instruction.rs [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+const ADC_SET: [Instruction; 8] = [
+    Instruction{name: "ADC", opcode: 0x69, mode: Mode::Immediate,   len: 2},
+    Instruction{name: "ADC", opcode: 0x65, mode: Mode::ZeroPage,    len: 2},
+    Instruction{name: "ADC", opcode: 0x75, mode: Mode::ZeroPageX,   len: 2},
+    Instruction{name: "ADC", opcode: 0x6D, mode: Mode::Absolute,    len: 3},
+    Instruction{name: "ADC", opcode: 0x7D, mode: Mode::AbsoluteX,   len: 3},
+    Instruction{name: "ADC", opcode: 0x79, mode: Mode::AbsoluteY,   len: 3},
+    Instruction{name: "ADC", opcode: 0x61, mode: Mode::IndirectX,   len: 2},
+    Instruction{name: "ADC", opcode: 0x71, mode: Mode::IndirectY,   len: 2},
+];
```
Comment (wiktorwieclaw) [inline src/instruction.rs:37]: how are the instructions categorized? sometimes I see that they have the same `name`, but sometimes it's not the case. Are the categories defined in the spec?

File: src/instruction.rs [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+        let mut v = Vec::new();
+        v.push(&BRK);
+        v.push(&JSR);
+        v.push(&NOP);
+        v.push(&RTI);
+        v.push(&RTS);
+        for instruction in ADC_SET.iter()
+        {
+            v.push(instruction);
+        }
```
Comment (wiktorwieclaw) [inline src/instruction.rs:269]: i think that its a lot of for loops :D https://doc.rust-lang.org/std/vec/struct.Vec.html#method.extend_from_slice

File: src/instruction.rs [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+        m
+    };
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{INSTRUCTION_MAP, Mode, Instruction};
+
+    fn get_lenght_from_mode(mode: &Mode) -> u8
+    {
```
Comment (wiktorwieclaw) [inline src/instruction.rs:389]: we need to use a formatter. For now I'd like to ask you to use rustfmt manually, but I guess we could setup the CI that does this automatically later on

File: src/instruction.rs [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+        {
+            v.push(instruction);
+        }
+        v
+    };
+
+    pub static ref INSTRUCTION_MAP : std::collections::HashMap<u8, &'static Instruction> = {
+        let mut m = std::collections::HashMap::<u8,&'static Instruction>::new();
+        for instruction in INSTRUCTIONS.iter() {
+            if let Some(_) = m.insert(instruction.opcode, instruction) {
```
Comment (wiktorwieclaw) [inline src/instruction.rs:376]: using `is_some()` would make it a little shorter :)

**Comment by KOomaXX** [comment]
I was thinking the same. But i wanted your opinion 😅

Le sam. 5 mars 2022 à 14:13, Wiktor Więcław ***@***.***> a
écrit :

File: src/instruction.rs (hunk 1 above)
Comment (wiktorwieclaw) [reply src/instruction.rs:37]: ok I stopped being lazy and looked it up and I see that it's from the spec :D, what I find weird is that some of them are not defined in terms of sets (e.g. BRK), I think I'd rather have single item arrays just for consistency,,, but it doesnt really matter lol

File: src/instruction.rs (hunk 3 above)
Comment (KOomaXX) [reply src/instruction.rs:389]: ok

File: src/instruction.rs (hunk 2 above)
Comment (KOomaXX) [reply src/instruction.rs:269]: I'll change it

File: src/instruction.rs (hunk 1 above)
Comment (KOomaXX) [reply src/instruction.rs:37]: I can do that for sure. Np

**Pull request merged**